### DISCUSSION TITLE
Usando switch case

### DISCUSSION THREAD

**Discussion by YasmimVieira** [frontendbr/forum#2056] [Perguntas e Respostas]
Galera, eu to fazendo um exercicio da faculdade que diz o seguinte:

/**
 * Função que recebe uma operação em formato de texto e devolve o resultado desta operação com os dois números passados,
 * na ordem em que eles foram passados. (Use as funções do Math se necessário)
 * 
 * As operações são identificadas da seguinte maneira:
 *  - A: Adição - Soma numero1 com numero2
 *  - S: Subtração - Subtrai numero2 do numero1
 *  - M: Multiplicação - Multiplica numero1 com numero2
 *  - D: Divisão - Divide numero1 pelo numero2
 *  - P: Potência - Eleva o numero1 pelo numero2
 * 
 * Verifique onde é possível fazer a operação (cuidado com divisão por zero e raízes pares de números negativos). 
 * Nos casos em que não for definido - divisão por zero, por exemplo - retornar NaN (not-a-number).
 * 
 * Se a operação passada não for uma das letras acima, retornar undefined.
 
 
 
 Aí beleza, como uma boa pratica eu usei o switch case por serem vários casos, porém está dando errado, mas quando uso o else if da certo. Podem me ajudar a identificar oq esta errado no meu switch case?

function operacoesBasicas(operacao, numero1, numero2) {
    let resultado

    switch(operacao) {
        case operacao === 'A':
          return resultado = numero1 + numero2;
        case operacao === 'S':
          return resultado = numero1 - numero2
        case operacao === 'M':
          return resultado = numero1 * numero2;
        case operacao === 'D':
          return resultado = numero1 / numero2;
        case operacao === 'P':
          return resultado = numero1 ** numero2;
        case 'D' / 0:
          return resultado = NaN;
        default:
          return resultado;
    }
}

**Comment by joaom00** [comment]
Acredito que seja a falta do break no final de cada case

**Comment by gabriel-hahn** [comment]
Olá @YasmimVieira,

Acredito que o problema é que você está retornando uma operação de associação de valor para o `resultado`. Há algumas maneiras de corrigir:

1 - Retornar o valor diretamente, por exemplo: `return numero1 / numero2` (retorna o valor da divisão) ao invés de `return resultado = numero1 / numero2` (da erro, já que você está fazendo uma associação, não deveria ter um `return` aqui).

Ex:

```
function operacoesBasicas(operacao, numero1, numero2) {
  switch(operacao) {
      case 'A':
        return numero1 + numero2;
      case 'S':
        return numero1 - numero2
      case 'M':
        return numero1 * numero2;
      case 'D':
        return numero1 / numero2;
      case 'P':
        return numero1 ** numero2;
      case 'D' / 0:
        return NaN; // Só uma observação, aqui acho que vale a pena validar a divisão antes de entrar no switch, pois sua validação de NaN deveria ser feito nos cálculos envolvendo o numero1 e numero2, não o case em si, mas dai é outro caso hehe :)
      default:
        return resultado;
  }
}
```

2 - Retirar os `return`, deixando somente `resultado = numero1 / numero2` e utilizar o `break` após cada caso (o `return` faz com que não seja necessário utilizar o `break`, pois o `return` já quebra o fluxo do `switch`, retornando um valor). E no final da sua função, fazer um `return resultado`.

Ex:

```
function operacoesBasicas(operacao, numero1, numero2) {
  let resultado;
  
  switch(operacao) {
      case 'A':
       resultado = numero1 + numero2;
       break;
      case 'S':
       resultado = numero1 - numero2
       break;
      case 'M':
       resultado = numero1 * numero2;
       break;
      case 'D':
       resultado = numero1 / numero2;
       break;
      case 'P':
       resultado = numero1 ** numero2;
       break;
      default:
        return resultado;
  }
  
  return resultado;
 }
```

Espero ter ajudado e que tenha ficado claro hehe abraço!

**Reply by dziul** [reply]
@gabriel-hahn , só uma correção, no `case` apenas adicionar o valor que quer comparar. Assim, sua sugestão ficaria:

```js
//js
function operacoesBasicas(operacao, numero1, numero2) {
  let resultado;
  
  switch(operacao) {
      case 'A':
       resultado = numero1 + numero2;
       break;
      case 'S':
       resultado = numero1 - numero2
       break;
      case 'M':
       resultado = numero1 * numero2;
       break;
      case 'D':
       resultado = numero1 / numero2;
       break;
      case 'P':
       resultado = numero1 ** numero2;
       break;
      case 'D':
       resultado = NaN;
       break;
  }  
  return resultado;
 }
```

ou mesmo o que @YasmimVieira estava tentado fazer, funciona com correções:

```js
//js
function operacoesBasicas(operacao, numero1, numero2) {
  
  switch(operacao) {
      case 'A':
           return numero1 + numero2;
      case 'S':
           return  numero1 - numero2
      case 'M':
           return numero1 * numero2;
      case 'D':
           return numero1 / numero2;
      case 'P':
           return numero1 ** numero2;
      case 'D':
           return  NaN;
      default: 
          return undefined;
    }  
  
}
```


[mais sobre switch](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Statements/switch)

**Reply by gabriel-hahn** [reply]
Boa @dziul, verdade, esse detalhe passou batido. Obrigado 👌🏼 
Corrigi o `case` da minha sugestão conforme você comentou.
